src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -0,0 +1,167 @@
+      // operators (e.g., outer joins) might have added NULLs to a primary key column. For now, we take the safe route
+      // and ignore all cases where group-by columns are nullable.
+      if (expression->is_nullable_on_lqp(aggregate_node)) {
+        continue;
+      }
+
+      const auto& column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression);
+      if (!column_expression) {
+        // In case the group-by column is not a column expression (e.g., grouping by `a+1`), we take the safe route and
+        // continue even though constraints often hold for such simple expression.
+        continue;
+      }
+
+      const auto& stored_table_node =
+          std::dynamic_pointer_cast<const StoredTableNode>(column_expression->column_reference.original_node());
+      // If column is not a physical column skip
+      if (!stored_table_node) continue;
+
+      const auto column_id = column_expression->column_reference.original_column_id();
+      group_by_columns_per_table[stored_table_node].insert(column_id);
+    }
+
+    // Store copy of the aggregate's column expressions to later check if this order is the query's final column order
+    const auto initial_aggregate_column_expressions = aggregate_node.column_expressions();
+    auto group_by_list_changed = false;
+
+    // Main loop. Iterate over the tables and its group-by columns, get table constraints, and try to reduce for each
+    // constraint.
+    for (const auto& [stored_table_node, group_by_columns] : group_by_columns_per_table) {
+      auto unique_columns = std::set<ColumnID>();
+
+      const auto& table = Hyrise::get().storage_manager.get_table(stored_table_node->table_name);
+      const auto& table_constraints = table->get_soft_unique_constraints();
+      if (table_constraints.empty()) {
+        // early exit for current table if no constraints are set
+        continue;
+      }
+
+      // Gather sizes of table constraints (i.e., the number of columns) to start with the shortest constraint (either
+      // unique or primary key) in hope that the shorter one will later form the group-by clause.
+      std::vector<std::pair<size_t, size_t>> constraints_position_and_size;
+      constraints_position_and_size.reserve(table_constraints.size());
+      auto constraint_id = size_t{0};
+      for (const auto& table_constraint : table_constraints) {
+        constraints_position_and_size.emplace_back(constraint_id, table_constraint.columns.size());
+        ++constraint_id;
+      }
+      std::sort(
+          constraints_position_and_size.begin(), constraints_position_and_size.end(),
+          [](const auto& constraint_1, const auto& constraint_2) { return constraint_1.second < constraint_2.second; });
+
+      // Try to reduce the group-by list one constraint at a time, starting with the shortest constraint. As soon as
+      // one reduction took place, we can ignore the remaining constraints.
```

Comment:
Hmm. Yeah, I think I had a wrong assumption about our constraints here. Will need to think about it.

Reply:
Just to be sure that I am not making a mistake here. Can there be a situation where we can gain an advantage when handling multiple constraints?
I thought that the shortest constraint should always be sufficient (not thinking about data types such as strings, which would actually make sense .. hmm) to minimise the group-by list. Then I thought that might not be the case, but then again I think it could still be the case.

Reply:
I haven't given this too much thought. You are probably correct, but I suggest to add an explanation to the claim that "we can ignore the remaining constraints".